src/Exception.php

```diff
@@ -85,82 +102,7 @@ public function getMyTrace()
      */
     public function getColorfulText()
     {
-        $output = "\033[1;31m--[ Agile Toolkit Exception ]---------------------------\n";
-        $output .= get_class($this).": \033[47m".$this->getMessage()."\033[0;31m".
-            ($this->getCode() ? ' [code: '.$this->getCode().']' : '');
-
-        foreach ($this->params as $key => $val) {
-            $key = str_pad($key, 19, ' ', STR_PAD_LEFT);
-            $output .= "\n".$key.': '.$this->toString($val);
-        }
-
-        $output .= "\n\033[0mStack Trace: ";
-
-        $in_atk = true;
-        $escape_frame = false;
-
-        foreach ($this->getMyTrace() as $call) {
-            if (!isset($call['file'])) {
-                $call['file'] = '';
-            } elseif (
-                $in_atk &&
-                strpos($call['file'], '/data/src/') === false &&
-                strpos($call['file'], '/core/src/') === false &&
-                strpos($call['file'], '/dsql/src/') === false
-            ) {
-                $escape_frame = true;
-                $in_atk = false;
-            }
-
-            $file = str_pad(substr($call['file'], -40), 40, ' ', STR_PAD_LEFT);
-
-            $line = str_pad(@$call['line'], 4, ' ', STR_PAD_LEFT);
-
-            $output .= "\n\033[0;34m".$file."\033[0m";
-            $output .= ":\033[0;31m".$line."\033[0m";
-
-            if (isset($call['object'])) {
-                $name = (!isset($call['object']->name)) ? get_class($call['object']) : $call['object']->name;
-                $output .= " - \033[0;32m".$name."\033[0m";
-            }
-
-            $output .= " \033[0;32m";
-
-            if (isset($call['class'])) {
-                $output .= $call['class'].'::';
-            }
-
-            if ($escape_frame) {
-                $output .= "\033[0,31m".$call['function'];
-                $escape_frame = false;
-
-                $args = [];
-                foreach ($call['args'] as $arg) {
-                    $args[] = $this->toString($arg);
-                }
-
-                $output .= "\n".str_repeat(' ', 20)."\033[0,31m(".implode(', ', $args);
-            } else {
-                $output .= "\033[0,33m".$call['function'].'(';
-            }
-
-            $output .= ')';
-        }
-
-        if ($p = $this->getPrevious()) {
-            $output .= "\n\033[0mCaused by Previous Exception:\n";
-            $output .= "\033[1;31m".get_class($p).': '.$p->getMessage()."\033[0;31m".
-                ($p->getCode() ? ' [code: '.$p->getCode().']' : '');
-            if ($p instanceof self) {
-                $output .= "\n+".str_replace("\n", "\n| ", $p->getColorfulText());
-            }
-        }
-
-        // next print params
-
-        $output .= "\n\033[1;31m-------------------------------------------------------\n";
-
-        return $output."\033[0m";
+        return (string) new Console($this);
```

Comment:
possibly pass translator inside this.. ?

Reply:
how about default callback here if exception is occured inside?

Reply:
App wouldn't really be available in here though, would it? So the singleton is the only option left.